### PR TITLE
Login Ready.

### PR DESCRIPTION
**Pull request by asolano-pernix**
login

### CODE REVIEW

File: app/components/layout/nav/nav-tpl.html [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
+<nav ng-controller="navController" class="navbar navbar-default"></nav><!-- Avoid padding top on main content -->
 <nav class="navbar navbar-default navbar-fixed-top">
   <div class="container">
     <div class="navbar-header page-scroll">
       <a class="navbar-brand" href="#"><img src="images/nav_logo.png" alt="Budget"></a>
     </div>
     <ul class="nav navbar-nav navbar-right pull-right">
       <li class="page-scroll navbar-options">
-        <a ui-sref="login">Cerrar Sesión</a>
+        <a ui-sref="login" ng-mousedown="vm.close()">Cerrar Sesión</a>
```
Comment (Emmurillo) [inline app/components/layout/nav/nav-tpl.html]: `vm.logout()` would be a better name

File: app/components/login/auth.service.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    .module('app.core')
+    .factory('AuthService', AuthService);
+
+  /* @ngInject */
+  function AuthService($state) {
+
+    var service = {
+
+      login:login,
+      verify_login:verify_login,
```
Comment (Emmurillo) [inline app/components/login/auth.service.js]: `camelCase` instead of `snake_case`

File: app/components/layout/nav/nav.controller.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+(function() {
+  'use strict';
+
+  angular
+    .module('app.login')
+    .controller('navController',navController);
+
+  /* @ngInject */
+  function navController(AuthService) {
```
Comment (johnnyxu-pernix) [inline app/components/layout/nav/nav.controller.js]: The controller name should start with *uppercase*

File: app/components/login/auth.service.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      logout:logout
+    };
+
+    return service;
+
+    function login(email,pass) {
+      firebase.auth().signInWithEmailAndPassword(email, pass).then(function () {
+          $state.go('home.listReserves');
+      }).catch(function() {
+        window.alert("Error de autentificación");
```
Comment (Emmurillo) [inline app/components/login/auth.service.js]: Please include [toastr](https://github.com/Foxandxss/angular-toastr) or something similar to show this kind of messages.

File: app/components/login/auth.service.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      });
+
+
+    }
+
+
+    function logout(){
+      firebase.auth().signOut().then(function() {
+        $state.go("login");
+      }, function(error) {
```
Comment (Emmurillo) [inline app/components/login/auth.service.js]: If this is doing nothing, then remove it.

File: app/components/login/login.controller.js [hunk 6]
```diff
@@ -6,16 +6,22 @@
     .controller('LoginController', LoginController);
 
   /* @ngInject */
-  function LoginController() {
-
+  function LoginController(AuthService) {
     var vm = this;
+    vm.authentication_user = authentication_user;
```
Comment (Emmurillo) [inline app/components/login/login.controller.js]: Same here, camelCase

File: app/components/login/login.html [hunk 7]
```diff
@@ -1,10 +1,21 @@
 <!DOCTYPE html>
```
Comment (Emmurillo) [inline app/components/login/login.html]: no need to create a whole html markup here, it's just a template.

File: app/components/login/login.html [hunk 8]
```diff
@@ -1,10 +1,21 @@
...
 <head>
+  <link href="../../styles/login.css" rel="stylesheet">
   <meta charset="UTF-8">
   <title>Login</title>
 </head>
-<body>
-  <h1>Login</h1>
+<body ng-controller="LoginController">
+  <section class="login-form-wrap">
+    <IMG class="icon"  src="../../images/bgt_hrz_rgb_pos%20copy.png"/>
```
Comment (johnnyxu-pernix) [inline app/components/login/login.html]: Why is the img tag in uppercase?

Comment (johnnyxu-pernix) [inline app/components/login/login.html]: Change the image path to `/images/image_name`

Comment (johnnyxu-pernix) [inline app/components/login/login.html]: `budgetLogo.png` would be a better file name.

File: app/components/login/login.html [hunk 9]
```diff
@@ -1,10 +1,21 @@
...
+      <div class="login-form-general">
+        <form class="login-form">
+          <input type="email" ng-model= "vm.user.email" placeholder="Correo Electronico"  id="username" required autofocus>
```
Comment (johnnyxu-pernix) [inline app/components/login/login.html]: `Electrónico`

File: app/components/login/login.html [hunk 10]
```diff
@@ -1,10 +1,21 @@
...
+          <input type="password"    ng-model="vm.user.pass" placeholder="Contraseña"  id="userPassword" required >
```
Comment (johnnyxu-pernix) [inline app/components/login/login.html]: delete the uncecesary whitespaces

File: app/components/login/login.html [hunk 11]
```diff
@@ -1,10 +1,21 @@
...
+          <input type="submit" value="Entrar"   ng-click="vm.authentication_user()"></ino>
```
Comment (johnnyxu-pernix) [inline app/components/login/login.html]: same here

Comment (johnnyxu-pernix) [inline app/components/login/login.html]: Use camelCase for the function name.

File: app/components/layout/nav/nav-tpl.html (hunk 1 above)
Comment (johnnyxu-pernix) [inline app/components/layout/nav/nav-tpl.html]: @Emmurillo would it be better to use `ng-click`?

**Comment by Emmurillo** [comment]
@johnnyxu-pernix is that grey background color supposed to be that way? It looks weird to me

File: app/components/layout/nav/nav-tpl.html [hunk 12]
```diff
@@ -1,12 +1,12 @@
-<nav class="navbar navbar-default"></nav><!-- Avoid padding top on main content -->
+<nav ng-controller="navController" class="navbar navbar-default"></nav><!-- Avoid padding top on main content -->
```
Comment (Emmurillo) [inline app/components/layout/nav/nav-tpl.html]: no need to add `ng-controller`, it's binding it within the `component` declaration.

File: app/components/layout/nav/nav-tpl.html (hunk 1 above)
Comment (Emmurillo) [inline app/components/layout/nav/nav-tpl.html]: Use `ng-click` instead of `ng-mousedown`

File: app/components/layout/nav/nav.controller.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    var vm = this;
+
+    vm.close = close;
+
+
+    activate();
+
+    function activate() {
+      vm.user = {};
```
Comment (Emmurillo) [inline app/components/layout/nav/nav.controller.js]: Why to add this empty user here?

Comment (Emmurillo) [inline app/components/layout/nav/nav.controller.js]: Remove it and therefore, the activate function

File: app/components/login/auth.service.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+      login:login,
+      verify_login:verify_login,
+      logout:logout
+    };
+
+    return service;
+
+    function login(email,pass) {
+      firebase.auth().signInWithEmailAndPassword(email, pass).then(function () {
+          $state.go('home.listReserves');
```
Comment (Emmurillo) [inline app/components/login/auth.service.js]: You shouldn't be handling UI interaction and state transitions within services.

File: app/components/login/auth.service.js [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+      firebase.auth().onAuthStateChanged(function(firebaseUser) {
+        if (firebaseUser) {
+          $state.go('home.listReserves');
+        } else {
+          $state.go("login");
+        }
+      });
+
+
+    }
```
Comment (johnnyxu-pernix) [inline app/components/login/auth.service.js]: Remove the unnecessary space

File: app/components/login/login.controller.js [hunk 16]
```diff
@@ -6,16 +6,22 @@
...
+    vm.close = close;
+
 
     activate();
 
     function activate() {
 
+      vm.user = {};
+      AuthService.verify_login();
     }
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: Delete the unnecessary  space

File: app/components/login/login.controller.js [hunk 17]
```diff
@@ -6,16 +6,22 @@
...
 
-  }
+    function authentication_user() {
+      AuthService.login(vm.user.email,vm.user.pass);
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: Add space between the `,` and `vm.user.pass`

**Comment by johnnyxu-pernix** [comment]
@Emmurillo actually yes, we're using this mockup:
web 1920 6

**Comment by johnnyxu-pernix** [comment]
@asolano-pernix use bootstrap classes instead of creating your own. If you need any help with bootstrap you can ask either your mentor, me or anyone in the team.

File: app/components/login/auth.service.js [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+
+
+    function verify_login(){
+      firebase.auth().onAuthStateChanged(function(firebaseUser) {
+        if (firebaseUser) {
+          $state.go('home.listReserves');
+        } else {
+          $state.go("login");
+        }
```
Comment (johnnyxu-pernix) [inline app/components/login/auth.service.js]: Is the else necessary? Are we not in the login here?

File: app/components/login/login.controller.js (hunk 17 above)
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: `camelCase`

File: app/components/login/login.html [hunk 19]
```diff
@@ -1,10 +1,21 @@
 <!DOCTYPE html>
 <html lang="en">
 <head>
+  <link href="../../styles/login.css" rel="stylesheet">
   <meta charset="UTF-8">
   <title>Login</title>
 </head>
-<body>
-  <h1>Login</h1>
+<body ng-controller="LoginController">
```
Comment (johnnyxu-pernix) [inline app/components/login/login.html]: As @Emmurillo said 

> no need to add ng-controller, it's binding it within the component declaration.

**Comment by johnnyxu-pernix** [comment]
@asolano-pernix tip: run `gulp review` to check the syntax.

File: app/components/login/login.html [hunk 20]
```diff
@@ -1,10 +1,20 @@
...
-  <h1>Login</h1>
+  <section class="login-form-wrap">
+    <img class="icon" src="images/budgetLogo.png"/>
+      <div class="login-form-general">
+        <form class="login-form">
+          <input type="email" ng-model= "vm.user.email" placeholder="Correo Electrónico"  id="username" required autofocus>
+          <input type="password" ng-model="vm.user.pass" placeholder="Contraseña"  id="userPassword" required >
+          <input type="submit" value="Entrar" ng-click="vm.authenticationUser()"></ino>
+        </form>
+  </div>
```
Comment (johnnyxu-pernix) [inline app/components/login/login.html]: Fix identation

File: app/components/app.config.js [hunk 21]
```diff
@@ -9,5 +9,6 @@
   function configuration($locationProvider) {
     $locationProvider.html5Mode(true);
   }
+
```
Comment (johnnyxu-pernix) [inline app/components/app.config.js]: Delete this line

File: app/components/layout/nav/nav-tpl.html [hunk 22]
```diff
@@ -6,7 +6,7 @@
     </div>
     <ul class="nav navbar-nav navbar-right pull-right">
       <li class="page-scroll navbar-options">
-        <a ui-sref="login">Cerrar Sesión</a>
+        <a ui-sref="login" ng-mousedown="vm.logout()">Cerrar Sesión</a>
```
Comment (johnnyxu-pernix) [inline app/components/layout/nav/nav-tpl.html]: The redirection to login should be done in the function.

File: app/components/layout/nav/nav.controller.js [hunk 23]
```diff
@@ -0,0 +1,19 @@
+(function() {
+  'use strict';
+
+  angular
+    .module('app.login')
+    .controller('NavController',NavController);
```
Comment (johnnyxu-pernix) [inline app/components/layout/nav/nav.controller.js]: Add a space after the `,`

File: app/components/layout/nav/nav.controller.js [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+
+  /* @ngInject */
+  function NavController(AuthService,$state) {
```
Comment (johnnyxu-pernix) [inline app/components/layout/nav/nav.controller.js]: Same here space after the `,`

File: app/components/layout/nav/nav.controller.js [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+    var vm = this;
+
+    vm.logout = logout;
+
+    function logout() {
+      AuthService.logout().then(function() {
+        $state.go("login");
+      })
```
Comment (johnnyxu-pernix) [inline app/components/layout/nav/nav.controller.js]: Create a function and call it in the then.

File: app/components/login/auth.service.js [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+    .module('app.core')
+    .factory('AuthService', AuthService);
+
+  /* @ngInject */
+  function AuthService() {
+    var service = {
+
+      login:login,
+      getState:getState,
+      logout:logout,
```
Comment (johnnyxu-pernix) [inline app/components/login/auth.service.js]: CamelCase

File: app/components/login/auth.service.js [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+
+  angular
+    .module('app.core')
+    .factory('AuthService', AuthService);
+
+  /* @ngInject */
+  function AuthService() {
+    var service = {
+
+      login:login,
```
Comment (johnnyxu-pernix) [inline app/components/login/auth.service.js]: camelCase

File: app/components/login/auth.service.js [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+      getState:getState,
+      logout:logout,
+    };
+
+    return service;
+
+    function login(email,pass) {
```
Comment (johnnyxu-pernix) [inline app/components/login/auth.service.js]: Add a space after the comma

File: app/components/login/auth.service.js [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+
+      return firebase.auth().signInWithEmailAndPassword(email, pass);
```
Comment (johnnyxu-pernix) [inline app/components/login/auth.service.js]: Unnecessary space here

File: app/components/login/auth.service.js [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+
+      login:login,
+      getState:getState,
+      logout:logout,
+    };
+
+    return service;
+
+    function login(email,pass) {
+
```
Comment (johnnyxu-pernix) [inline app/components/login/auth.service.js]: Unnecessary space here

File: app/components/login/auth.service.js [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+      return firebase.auth().signInWithEmailAndPassword(email, pass);
+    }
+
+    function getState(){
+      return firebase.auth();
+
```
Comment (johnnyxu-pernix) [inline app/components/login/auth.service.js]: Same here

File: app/components/login/login.controller.js [hunk 32]
```diff
@@ -6,16 +6,31 @@
     .controller('LoginController', LoginController);
 
   /* @ngInject */
-  function LoginController() {
-
+  function LoginController(AuthService,$state,toastr) {
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: Add space after the `,`

File: app/components/login/login.controller.js [hunk 33]
```diff
@@ -6,16 +6,31 @@
...
+      vm.user = {};
+     verifyLogin();
+    }
+    
+    function verifyLogin() {
+      AuthService.getState().onAuthStateChanged(function (user) {
+        if(user){
+          $state.go('home.listReserves');
+        }
+      })
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: Create a function and call it in the `onAuthStateChanged`

File: app/components/login/login.controller.js [hunk 34]
```diff
@@ -6,16 +6,31 @@
...
     }
 
-  }
+    function authenticationUser() {
+      AuthService.login(vm.user.email, vm.user.pass).then(function () {
+        $state.go('home.listReserves');
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: Same as above

File: app/components/login/login.controller.js [hunk 35]
```diff
@@ -6,16 +6,31 @@
...
+      }).catch(function() {
+        toastr.error('Usuario o contraseña incorrectos','Autentificación Incorrecta');
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: space after the `,`

File: app/index.html [hunk 36]
```diff
@@ -54,6 +57,12 @@
     <script src="components/users/delete/user.delete.controller.js"></script>
     <script src="components/users/list/users.list.controller.js"></script>
     <script src="components/users/show/user.show.controller.js"></script>
+    <script src="components/login/auth.service.js"></script>
+    <script src="components/layout/nav/nav.controller.js"></script>
+
+
+
+
     <!-- endinject -->
```
Comment (johnnyxu-pernix) [inline app/index.html]: Did you use the gulp task to inject the files?

**Comment by johnnyxu-pernix** [comment]
@asolano-pernix as mentioned before you should use Bootstrap components instead of creating your own classes.

File: app/components/login/auth.service.js (hunk 28 above)
Comment (johnnyxu-pernix) [inline app/components/login/auth.service.js]: I think `authenticate` would be a better name for the function.

File: app/components/login/login.controller.js [hunk 37]
```diff
@@ -6,16 +6,31 @@
...
-
+    vm.authenticationUser = authenticationUser;
+    vm.verifyLogin = verifyLogin;
     activate();
 
     function activate() {
-
+      vm.user = {};
+     verifyLogin();
+    }
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: Why are you calling verifyLogin() in the activate?

File: app/components/layout/nav/nav.controller.js [hunk 38]
```diff
@@ -0,0 +1,21 @@
...
+
+  /* @ngInject */
+  function NavController(authService, $state) {
+    var vm = this;
+    vm.RedirectToLogin=RedirectToLogin;
+
+    vm.logOut = logOut;
+
+    function logOut() {
+      authService.logOut().then(RedirectToLogin);
```
Comment (joserv2812) [inline app/components/layout/nav/nav.controller.js]: .then in a new line

File: app/components/login/login.controller.js [hunk 39]
```diff
@@ -6,16 +6,35 @@
...
+      authService.logIn(vm.user.email, vm.user.pass).then(RedirectToHome)
+        .catch(showError);
+    }
 
+    function showError() {
+      toastr.error('Usuario o contraseña incorrectos', 'Autentificación Incorrecta');
     }
 
-  }
+    function RedirectToHome(user) {
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: @asolano-pernix the functions should start with lowercase

File: app/components/layout/nav/nav.controller.js [hunk 40]
```diff
@@ -0,0 +1,21 @@
...
+    }
+    function RedirectToLogin(){
```
Comment (joserv2812) [inline app/components/layout/nav/nav.controller.js]: change to redirectToLogin

File: app/components/login/login.controller.js [hunk 41]
```diff
@@ -6,16 +6,35 @@
...
   /* @ngInject */
-  function LoginController() {
-
+  function LoginController(authService, $state, toastr) {
     var vm = this;
-
+    vm.authenticationUser = authenticationUser;
+    vm.verifyLogin = verifyLogin;
+    vm.RedirectToHome=RedirectToHome;
+    vm.showError= showError;
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: Add space before the `=`

File: app/styles/login.css [hunk 42]
```diff
@@ -0,0 +1,45 @@
...
+  width: 100%;
+  background-color: #D7D6D8;
+  overflow: hidden;
+}
+
+.form-input{
+  margin-top: 5%;
+
+
+}
```
Comment (johnnyxu-pernix) [inline app/styles/login.css]: Delete the unnecessary lines

File: app/components/login/login.controller.js (hunk 39 above)
Comment (joserv2812) [inline app/components/login/login.controller.js]: same here

File: app/components/login/login.html [hunk 43]
```diff
@@ -1,10 +1,22 @@
...
+      </div>
+  </div>
+</div>
+</div>
+
+
+
+
+
+
```
Comment (joserv2812) [inline app/components/login/login.html]: this lines?

File: app/components/layout/nav/nav.controller.js [hunk 44]
```diff
@@ -0,0 +1,22 @@
...
+  'use strict';
+
+  angular
+    .module('app.login')
+    .controller('NavController', NavController);
+
+  /* @ngInject */
+  function NavController(authService, $state) {
+    var vm = this;
+    vm.redirectToLogin=redirectToLogin;
```
Comment (johnnyxu-pernix) [inline app/components/layout/nav/nav.controller.js]: Add a space after and before the `=`

File: app/components/login/login.controller.js [hunk 45]
```diff
@@ -6,16 +6,34 @@
...
+      vm.user = {};
+    }
 
+    function verifyLogin() {
+      authService.getState().onAuthStateChanged(RedirectToHome);
     }
 
-  }
+    function authenticationUser() {
+      authService.logIn(vm.user.email, vm.user.pass).then(RedirectToHome)
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: Like @joserv2812 the `.then` should be in a new line

File: package.json [hunk 46]
```diff
@@ -4,7 +4,8 @@
   "dependencies": {
     "angular-confirm": "^1.2.5",
     "angular-ui-bootstrap": "1.3.3",
-    "gulp": "3.9.1"
+    "gulp": "3.9.1",
+    "toastr": "^2.1.2"
```
Comment (johnnyxu-pernix) [inline package.json]: @asolano-pernix the `toastr` is already installed with `bower`.  I don't know which one are you using, but you should only have one.

File: app/components/login/login.controller.js [hunk 47]
```diff
@@ -6,16 +6,34 @@
...
     function activate() {
+      vm.user = {};
+    }
 
+    function verifyLogin() {
+      authService.getState().onAuthStateChanged(RedirectToHome);
     }
 
-  }
+    function authenticationUser() {
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: The function name should be `authenticateUser`

File: app/components/login/login.controller.js [hunk 48]
```diff
@@ -6,16 +6,34 @@
...
+      authService.logIn(vm.user.email, vm.user.pass).then(RedirectToHome)
+        .catch(showError);
+    }
 
-})();
+    function showError() {
+      toastr.error('Usuario o contraseña incorrectos', 'Autentificación Incorrecta');
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: Typo `Autenticación`

File: app/components/login/login.controller.js [hunk 49]
```diff
@@ -6,16 +6,35 @@
...
+    vm.redirectToHome = redirectToHome;
+    vm.showError = showError;
     activate();
 
     function activate() {
+      vm.user = {};
+    }
 
+    function verifyLogin() {
+      authService.getState().onAuthStateChanged(RedirectToHome);
```
Comment (johnnyxu-pernix) [inline app/components/login/login.controller.js]: @asolano-pernix you didn't change the function name here, for the next time please test the app before uploading the changes.

**Pull request closed without merging**